## Module 9 - Case study 3

1. Load the data from “college.csv” that has attributes collected about private and public colleges for a particular year. We will try to predict the private/public status of the college from other attributes.

2. Use LabelEncoder to encode the target variable in to numerical form and split the data such that 20% of the data is set aside for testing.

3. Fit a linear svm from scikit learn and observe the accuracy.

[Hint: Use Linear SVC]

4. Preprocess the data using StandardScalar and fit the same model again and observe the change in accuracy.

[Hint: Refer to scikitlearn’s preprocessing methods]

5. Use scikit learn’s gridsearch to select the best hyperparameter for a non-linear SVM, identify the model with best score and its parameters.

[Hint: Refer to model_selection module of Scikit learn]

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

C:\Users\sreeraj.va\AppData\Local\Continuum\anaconda3\lib\site-packages\statsmodels\tools\_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [2]:
df = pd.read_csv(r'D:\E\Courses\Edureka\Assignments\Dataset\module9\College.csv')

In [3]:
print(df.shape)
df.head(2)

(777, 18)


Private  Apps  Accept  Enroll  Top10perc  Top25perc  F.Undergrad  \
0     Yes  1660    1232     721         23         52         2885   
1     Yes  2186    1924     512         16         29         2683   

   P.Undergrad  Outstate  Room.Board  Books  Personal  PhD  Terminal  \
0          537      7440        3300    450      2200   70        78   
1         1227     12280        6450    750      1500   29        30   

   S.F.Ratio  perc.alumni  Expend  Grad.Rate  
0       18.1           12    7041         60  
1       12.2           16   10527         56

In [4]:
df.isna().sum()

Private        0
Apps           0
Accept         0
Enroll         0
Top10perc      0
Top25perc      0
F.Undergrad    0
P.Undergrad    0
Outstate       0
Room.Board     0
Books          0
Personal       0
PhD            0
Terminal       0
S.F.Ratio      0
perc.alumni    0
Expend         0
Grad.Rate      0
dtype: int64

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 777 entries, 0 to 776
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Private      777 non-null    object 
 1   Apps         777 non-null    int64  
 2   Accept       777 non-null    int64  
 3   Enroll       777 non-null    int64  
 4   Top10perc    777 non-null    int64  
 5   Top25perc    777 non-null    int64  
 6   F.Undergrad  777 non-null    int64  
 7   P.Undergrad  777 non-null    int64  
 8   Outstate     777 non-null    int64  
 9   Room.Board   777 non-null    int64  
 10  Books        777 non-null    int64  
 11  Personal     777 non-null    int64  
 12  PhD          777 non-null    int64  
 13  Terminal     777 non-null    int64  
 14  S.F.Ratio    777 non-null    float64
 15  perc.alumni  777 non-null    int64  
 16  Expend       777 non-null    int64  
 17  Grad.Rate    777 non-null    int64  
dtypes: float64(1), int64(16), object(1)
memory usage: 

In [6]:
x = df.iloc[:,1:].values
y =  df.iloc[:,0:1].values

In [7]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y = le.fit_transform(y)

C:\Users\sreeraj.va\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\preprocessing\label.py:235: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [8]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2, random_state=0)

In [9]:
len(x_train), len(x_test), len(y_train), len(y_test)

(621, 156, 621, 156)

In [10]:
from sklearn.svm import SVC

svc_lin = SVC(kernel='linear')
svc_lin.fit(x_train, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='linear', max_iter=-1, probability=False, random_state=None,
    shrinking=True, tol=0.001, verbose=False)

In [11]:
y_pred = svc_lin.predict(x_test)

In [12]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

In [13]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.90      0.86        41
           1       0.96      0.93      0.95       115

    accuracy                           0.92       156
   macro avg       0.89      0.92      0.90       156
weighted avg       0.93      0.92      0.92       156



In [14]:
confusion_matrix(y_test, y_pred)

array([[ 37,   4],
       [  8, 107]], dtype=int64)

In [15]:
accuracy_score(y_test, y_pred)

0.9230769230769231

###### applying standardscaler and again applying svm with linear kernel

In [16]:
from sklearn.preprocessing import StandardScaler

In [17]:
sc_x = StandardScaler()
x2 = sc_x.fit_transform(x)

In [18]:
x2_train, x2_test, y_train, y_test = train_test_split(x2, y, test_size=0.2, random_state=0)

In [19]:
svc_lin_sc = SVC(kernel='linear')
svc_lin_sc.fit(x2_train, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='linear', max_iter=-1, probability=False, random_state=None,
    shrinking=True, tol=0.001, verbose=False)

In [20]:
y_pred_sc = svc_lin_sc.predict(x2_test)

In [21]:
confusion_matrix(y_test, y_pred_sc)

array([[ 37,   4],
       [  6, 109]], dtype=int64)

In [22]:
accuracy_score(y_test, y_pred_sc)

0.9358974358974359

###### Accuract increased from 92.3% to 93.5% when applies standard scalar to the independant variables.

### Applying gridsearch to the model

In [23]:
from sklearn.model_selection import GridSearchCV

In [24]:
parameters = [{'C':[1,2,3,4,5], 'kernel':['sigmoid', 'rbf', 'linear', 'poly']}]

In [25]:
# applying to the model created without standardscaler

grid_search = GridSearchCV(estimator=svc_lin, param_grid=parameters,
                          scoring='accuracy', n_jobs=-1)
grid_search = grid_search.fit(x_train, y_train)
best_accuracy = grid_search.best_score_
best_parameters = grid_search.best_params_

C:\Users\sreeraj.va\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)


In [26]:
best_accuracy

0.9307568438003221

In [27]:
best_parameters

{'C': 1, 'kernel': 'linear'}

In [29]:
# applying to the model created after applying standardscaler

parameters2 = [{'kernel':['sigmoid', 'rbf', 'linear', 'poly']}]

grid_search_sc = GridSearchCV(estimator=svc_lin_sc, param_grid=parameters2,
                          scoring='accuracy', n_jobs=-1)
grid_search_sc = grid_search_sc.fit(x2_train, y_train)
best_accuracy_sc = grid_search_sc.best_score_
best_parameters_sc = grid_search_sc.best_params_

C:\Users\sreeraj.va\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)


In [30]:
best_accuracy_sc

0.9484702093397746

In [31]:
best_parameters_sc

{'kernel': 'linear'}

1. Lineal kernel is the best 
2. Applying  standardscaler will increase the accuracy of prediction in this case.